Add ROS blogpost (#276)
conan-io/conan-io.github.io: _posts/2024-11-28-Enhancing-ros-builds-with-Conan.markdown

---
layout: post
comments: false
title: "Enhancing ROS development with Conan packages"
meta_title: "Conan receives new integration for ROS to enhance robotics development."
description: "Conan has intorduced a new integration for ROS (Robot Operating System) to incorporate packages from Conan Center to any robotics development in C++."
keywords: "conan, ros, ros2, robots, robotics, c++, packages, libraries"
---

> When we think about developments in C or C++, we immediately associate them with programming for industrial systems and embedded devices.

Reviewer: ```suggestion
When we think about development in C or C++, we immediately associate it with programming for industrial systems and embedded devices.
```

Indeed, it is in this area where the language is extensively used, and this inevitably includes its application in robotics.

> In this context, there is no doubt that ROS (Robot Operating System) is the most well-known framework and probably one of the most widely used. It enables the integration of different hardware components into a single system, allowing them to function seamlessly together. This makes development in robotics more efficient, as it allows for the reuse of each component's software across projects for various robotic applications.

Reviewer: A link here for the ROS homepage might be nice

Author: yes, done. Thank you!

> This is why the Conan team is thrilled to present this integration of Conan with ROS. Many members of our team come from industrial engineering are passionate about this area, so being able to contribute to robotics through Conan makes us very happy.

Reviewer: ```suggestion
This is why the Conan team is thrilled to present this integration of Conan with ROS. Many members of our team come from industrial engineering and we are passionate about this area, so being able to contribute to robotics through Conan makes us very happy.
```

In this post, we will discuss the different components that make up ROS development, how they work, and how Conan integrates into these projects to manage third-party libraries.

## A Quick Primer on ROS2 Packages and Workspaces

Let's start with a brief overview for those less familiar with the ROS2 ecosystem.

ROS organizes code into packages, each containing a package.xml with some metadata, executables, libraries, or other resources like launch files. These packages are typically managed within a **workspace**, allowing developers to build and run their robot applications. The key components of the workflow are:

- **CMake for builds**: ROS2 heavily relies on CMake as its build system for C/C++ packages, using CMakeLists.txt files to define the build logic.

> - **Ament build tools**: The Ament build tools are a framework of CMake macros and functions that standardize and simplify package-level tasks

Reviewer: ```suggestion
- **Ament build tools**: The Ament build tools is a framework of CMake macros and functions that standardize and simplify package-level tasks
```

  such as linking dependencies, installation and export of artifacts or integration with ROS-specific tools like `rosidl`.

> - **Colcon**: It’s the main build tool that orchestrates the build of multiple

Reviewer: ```suggestion
- **Colcon**: It's the main build tool that orchestrates the build of multiple
```

```txt
cmake_minimum_required(VERSION 3.8)
project(navigation_package)

if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
  add_compile_options(-Wall -Wextra -Wpedantic)
endif()

# ROS dependencies
find_package(ament_cmake REQUIRED)
find_package(rclcpp REQUIRED)
find_package(rclcpp_action REQUIRED)
find_package(nav2_msgs REQUIRED)

# Conan dependencies
find_package(yaml-cpp REQUIRED)

add_executable(navigator src/main.cpp)

target_compile_features(navigator PUBLIC c_std_99 cxx_std_17)  # Require C99 and C++17
ament_target_dependencies(navigator rclcpp rclcpp_action nav2_msgs yaml-cpp)

install(
  TARGETS navigator
  DESTINATION lib/${PROJECT_NAME}
)

ament_package()
```

---
**Note**

In the case that we have to propagate the Conan packages as transitive dependencies for other ROS packages that depend on this one (in the case that navigation_package was a library):

  Other ROS Package → navigation_package → yaml-cpp Conan Package

We can use the ament helper `ament_export_dependencies()` to export the Conan targets as we would do with a normal ROS package. You can read more about it in our documentation: https://docs.conan.io/2/integration (LINK)

---

> Now we would install the ``yaml-cpp`` Conan package like so:

Reviewer: ```suggestion
Now we install the ``yaml-cpp`` Conan package like so:
```

```sh
$ cd navigation_ws
$ conan install navigation_package/conanfile.txt --build missing --output-folder install/conan
...
```

With this install command, Conan has performed some actions:
1. Search for packages in Conan Center, the central repository where OSS packages are contributed, that are suitable for your configuration.
2. Download the packages into the Conan cache locally in your machine.
3. Generate the environment and CMake files needed for your ROS project in the install/conan folder.

> Finally, let’s add the code for our node to the _main.cpp_ file:

Reviewer: ```suggestion
Finally, let's add the code for our node to the _main.cpp_ file:
```

First, we create a workspace `navigation_ws` folder, set the environment of your ROS installation, and create a package:

```sh
$ mkdir /home/navigation_ws && cd /home/navigation_ws
$ source /opt/ros/humble/setup.bash
$ ros2 pkg create --build-type ament_cmake --node-name navigator navigation_package
```

These are the files that should be in your workspace. As indicated in this post, you can inspect them to understand the contents.

```txt
navigation_ws/
  navigation_package/
    CMakeLists.txt
    package.xml
    src/
      navigator.cpp
```

Now we can invoke ``colcon`` to perform the build

```sh
$ colcon build --packages-select navigation_package
Starting >>> navigation_package
Finished <<< navigation_package [1.21s]

Summary: 1 package finished [1.60s]
```

Finally, before executing the binary, we have to set the environment for the executable (so it is able to locate any shared library that it may
have) and then we execute it:

```sh
$ source install/setup.bash
$ ros2 run navigation_package navigator
hello world navigation_package package
```

> ## An example of using ROS with Conan

Reviewer 2: I think this fits more what we are doing, if not it looks like this is another example, but it's the extension of the previous one
```suggestion
## Integrating Conan Dependencies into the ROS Example
```

            return false;
        }
    }

    void sendAllGoals() {
        for (const auto &location : locations_) {
            RCLCPP_INFO(this->get_logger(), "Sending goal to %s: (%.2f, %.2f)", location.name.c_str(), location.x, location.y);
            auto goal_msg = NavigateToPose::Goal();
            goal_msg.pose.header.frame_id = "map";
            goal_msg.pose.header.stamp = this->now();
            goal_msg.pose.pose.position.x = location.x;
            goal_msg.pose.pose.position.y = location.y;
            goal_msg.pose.pose.orientation.w = 1.0;
            action_client_->async_send_goal(goal_msg);
        }
        RCLCPP_INFO(this->get_logger(), "All goals have been sent.");
    }
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    if (argc < 2) {
        RCLCPP_ERROR(rclcpp::get_logger("yaml_navigation_node"), "Usage: yaml_navigation_node <yaml_file_path>");
        return 1;
    }
    std::string yaml_file_path = argv[1];
    std::make_shared<YamlNavigationNode>(yaml_file_path);
    rclcpp::shutdown();
    return 0;
}
```

The node will read the locations from the YAML file and send them as navigation goals for our robot.

> We can build our ROS package as usual, just source the conanrosenv.sh file before. This will set the environment so CMake can find the files generated by Conan.

Reviewer 2: ```suggestion
We can build our ROS package as usual, just source the `conanrosenv.sh` file before. This will set the environment so CMake can find the files generated by Conan.
```

_`navigation_ws/navigation_package/locations.yaml`_

```yaml
locations:
  - name: "Kitchen"
    x: 3.5
    y: 2.0
  - name: "Living Room"
    x: 1.0
    y: -1.0
  - name: "Bedroom"
    x: -2.0
    y: 1.5
```

And we will use the [``yaml-cpp`` library from Conan Center](https://conan.io/center/recipes/yaml-cpp). For that, we need to include a _conanfile.txt_ file next to the _CMakeLists.txt_ of our project:

_`navigation_ws/navigation_package/conanfile.txt`_

```txt
[requires]
yaml-cpp/0.8.0

[generators]
CMakeDeps
CMakeToolchain
ROSEnv
```

As you can see, we are listing our dependencies under the ``[requires]`` section and we are also adding the required generators, that _translate_ the information of the files in a package: libraries, headers, executables… into a file format that is suitable for the build system or the environment of the consumer):

- The **CMake ones** will generate the files so that the ``yaml-cpp`` package can be included in our project using ``find_package()``.

- The **ROSEnv generator** will create a shell script with the environment variables needed to perform the build.

> In the _CMakeLists.txt_, we will need to include the **ROS client libraries** (`find_package(rclcpp REQUIRED)`, and
> `find_package(rclcpp_action REQUIRED)`), the [ROS nav2_msgs](https://index.ros.org/p/nav2_msgs/) (`find_package(nav2_msgs REQUIRED)`) and also the [``yaml-cpp`` library from Conan](https://conan.io/center/recipes/yaml-cpp) (`find_package(yaml-cpp REQUIRED)`):

Reviewer 2: ```suggestion
In the _CMakeLists.txt_, we will need to include the **ROS client libraries** (`rclcpp` and `rclcpp_action`), the [ROS nav2_msgs](https://index.ros.org/p/nav2_msgs/), and the [``yaml-cpp`` library from Conan](https://conan.io/center/recipes/yaml-cpp). This is done using the `find_package` command for each dependency, as shown below:
```

endif()

# ROS dependencies
find_package(ament_cmake REQUIRED)
find_package(rclcpp REQUIRED)
find_package(rclcpp_action REQUIRED)
find_package(nav2_msgs REQUIRED)

# Conan dependencies
find_package(yaml-cpp REQUIRED)

add_executable(navigator src/navigator.cpp)

target_compile_features(navigator PUBLIC c_std_99 cxx_std_17)  # Require C99 and C++17
ament_target_dependencies(navigator rclcpp rclcpp_action nav2_msgs yaml-cpp)

install(TARGETS navigator
  DESTINATION lib/${PROJECT_NAME})

ament_package()
```

> **Note**:\
  In the case that we have to propagate the Conan packages as transitive dependencies for other ROS packages that depend on this one (in the case that ``navigation_package`` was a library):\
\
    ``Other ROS Package`` → ``navigation_package`` → ``yaml-cpp Conan Package``\
\
We can use the ament helper `ament_export_dependencies()` to export the Conan targets as we would do with a normal ROS package. You can read more about it in our documentation: <https://docs.conan.io/2/integrations/ros.html>

Now we install the ``yaml-cpp`` Conan package like so:

```sh
$ conan install navigation_package/conanfile.txt --build missing --output-folder install/conan

> ======== Input profiles ========

Reviewer 2: What about reducing a bit the output here? Just show the most relevant parts

Author: reduced

---
layout: post
comments: false

> title: "Enhancing ROS development with Conan packages"

Reviewer 3: Add "Robotics/Robots" and "C, C++" in the title?

Reviewer 3: Enhancing ROS robotics development with C,C++ Conan packages?

Author: ```suggestion
title: "Enhancing ROS robotics development with C/C++ Conan packages"
```

> meta_title: "Conan receives new integration for ROS to enhance robotics development."

Reviewer 2: ```suggestion
meta_title: "Conan introduces new integration for ROS to enhance robotics development."
```
receives sound a bit weird

> description: "Conan has intorduced a new integration for ROS (Robot Operating System) to incorporate packages from Conan Center to any robotics development in C++."

Reviewer 2: ```suggestion
description: "Conan has introduced a new integration for ROS (Robot Operating System) to incorporate packages from Conan Center to any robotics development in C++."
```

Indeed, it is in this area where the language is extensively used, and this inevitably includes its application in robotics.

In this context, there is no doubt that ROS (Robot Operating System) is the most well-known framework and probably one of the most widely used. It enables the integration of different hardware components into a single system, allowing them to function seamlessly together. This makes development in robotics more efficient, as it allows for the reuse of each component's software across projects for various robotic applications.

This is why the Conan team is thrilled to present this integration of Conan with ROS. Many members of our team come from industrial engineering are passionate about this area, so being able to contribute to robotics through Conan makes us very happy.
In this post, we will discuss the different components that make up ROS development, how they work, and how Conan integrates into these projects to manage third-party libraries.

## A Quick Primer on ROS2 Packages and Workspaces

Let's start with a brief overview for those less familiar with the ROS2 ecosystem.

> ROS organizes code into packages, each containing a package.xml with some metadata, executables, libraries, or other resources like launch files. These packages are typically managed within a **workspace**, allowing developers to build and run their robot applications. The key components of the workflow are:

Reviewer 2: ```suggestion
ROS organizes code into packages, each containing a `package.xml` with some metadata, executables, libraries, or other resources like launch files. These packages are typically managed within a **workspace**, allowing developers to build and run their robot applications. The key components of the workflow are:
```

```sh
$ mkdir /home/navigation_ws && cd /home/navigation_ws
$ source /opt/ros/humble/setup.bash
$ ros2 pkg create --build-type ament_cmake --node-name navigator navigation_package
```

These are the files that should be in your workspace. As indicated in this post, you can inspect them to understand the contents.

```txt
navigation_ws/
  navigation_package/
    CMakeLists.txt
    package.xml
    src/
      navigator.cpp
```

Now we can invoke ``colcon`` to perform the build

```sh
$ colcon build --packages-select navigation_package
Starting >>> navigation_package
Finished <<< navigation_package [1.21s]

Summary: 1 package finished [1.60s]
```

Finally, before executing the binary, we have to set the environment for the executable (so it is able to locate any shared library that it may
have) and then we execute it:

```sh
$ source install/setup.bash
$ ros2 run navigation_package navigator
hello world navigation_package package
```

## An example of using ROS with Conan

Let's say we want to include an external library using Conan. In this case, we would create a navigation node that sends locations goals from a

> yaml file to our mobile robot.

Reviewer 2: ```suggestion
`yaml` file to our mobile robot.
```

CMakeToolchain
ROSEnv
```

As you can see, we are listing our dependencies under the ``[requires]`` section and we are also adding the required generators, that _translate_ the information of the files in a package: libraries, headers, executables… into a file format that is suitable for the build system or the environment of the consumer):

- The **CMake ones** will generate the files so that the ``yaml-cpp`` package can be included in our project using ``find_package()``.

- The **ROSEnv generator** will create a shell script with the environment variables needed to perform the build.

In the _CMakeLists.txt_, we will need to include the **ROS client libraries**, the [ROS nav2_msgs](https://index.ros.org/p/nav2_msgs/) and also the [``yaml-cpp`` library from Conan](https://conan.io/center/recipes/yaml-cpp):

_`navigation_ws/navigation_package/CMakeLists.txt`_
```txt
cmake_minimum_required(VERSION 3.8)
project(navigation_package)

if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
  add_compile_options(-Wall -Wextra -Wpedantic)
endif()

# ROS dependencies
find_package(ament_cmake REQUIRED)
find_package(rclcpp REQUIRED)
find_package(rclcpp_action REQUIRED)

> find_package(nav2_msgs REQUIRED)

Reviewer 2: I think we should also mention these changes in the text? Event if that's just a sentence?

Author: the sentence is added before with links to the libraries. I have referenced the find_package there as well

conanfile.txt: CMakeToolchain generated: /navigation_ws/install/conan/CMakePresets.json
conanfile.txt: CMakeToolchain generated: /navigation_ws/navigation_package/CMakeUserPresets.json
conanfile.txt: Generator 'ROSEnv' calling 'generate()'
conanfile.txt: Generated ROSEnv Conan file: conanrosenv.sh
Use 'source /navigation_ws/install/conan/conanrosenv.sh' to set the ROSEnv Conan before 'colcon build'
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
Install finished successfully
```

With this install command, Conan has performed some actions:
1. **Search** for packages in [Conan Center](https://conan.io/center), the central repository where OSS packages are contributed, that are suitable for your configuration.
2. **Download** the packages into the Conan cache locally in your machine.
3. **Generate** the environment and CMake files needed for your ROS project in the _install/conan_ folder.

Finally, let’s add the code for our node to the _main.cpp_ file:

> **`navigation_ws/my_package/src/main.cpp`**

Reviewer 2: Maye too much code? what about putting the most relevant parts and for the rest `...`

> **`navigation_ws/my_package/src/main.cpp`**

Reviewer 2: If this comes from `ros2 pkg create --build-type ament_cmake --node-name navigator navigation_package` it will be called `navigator.cpp`

Author: ```suggestion
**`navigation_ws/my_package/src/navigator.cpp`**
```

---
layout: post
comments: false
title: "Enhancing ROS development with Conan packages"
meta_title: "Conan introduces new integration for ROS to enhance robotics development."
description: "Conan has introduced a new integration for ROS (Robot Operating System) to incorporate packages from Conan Center to any robotics development in C++."
keywords: "conan, ros, ros2, robots, robotics, c++, packages, libraries"
---

When we think about development in C or C++, we immediately associate it with programming for industrial systems and embedded devices.
Indeed, it is in this area where the language is extensively used, and this inevitably includes its application in robotics.

In this context, there is no doubt that [ROS (Robot Operating System)](https://ros.org/) is the most well-known framework and probably one of the most widely used. It enables the integration of different hardware components into a single system, allowing them to function seamlessly together. This makes development in robotics more efficient, as it allows for the reuse of each component's software across projects for various robotic applications.

> This is why the Conan team is thrilled to present this integration of Conan with ROS.Many members of our team come from industrial engineering and we are passionate about this area, so being able to contribute to robotics through Conan makes us very happy.

Reviewer 3: ```suggestion
This is why the Conan team is thrilled to present this integration of Conan with ROS. Many members of our team come from industrial engineering and we are passionate about this area, so being able to contribute to robotics through Conan makes us very happy.
```

> We know that `rosdep` is a well-known tool in the ROS ecosystem that helps with the task of installing dependencies into the system, and this somehow overlaps with the purpose of Conan as package manager. However, we also believe that Conan offers some advantages and could be introduced as an alternative for the developers that would like to benefit from the features of Conan an its collection of open-source packages.

Reviewer 3: ```suggestion
We know that `rosdep` is a well-known tool in the ROS ecosystem that helps with the task of installing dependencies into the system. However, managing some C and C++ dependencies at the system level is not always the best approach, and we believe that Conan offers some advantages and could be introduced as an alternative for the developers that would like to benefit from the features of Conan an its collection of open-source packages.
```

  add_compile_options(-Wall -Wextra -Wpedantic)
endif()

# ROS dependencies
find_package(ament_cmake REQUIRED)
find_package(rclcpp REQUIRED)
find_package(rclcpp_action REQUIRED)
find_package(nav2_msgs REQUIRED)

# Conan dependencies
find_package(yaml-cpp REQUIRED)

add_executable(navigator src/navigator.cpp)

target_compile_features(navigator PUBLIC c_std_99 cxx_std_17)  # Require C99 and C++17
ament_target_dependencies(navigator rclcpp rclcpp_action nav2_msgs yaml-cpp)

install(TARGETS navigator
  DESTINATION lib/${PROJECT_NAME})

ament_package()
```

> **Note**:\
  In the case that we have to propagate the Conan packages as transitive dependencies for other ROS packages that depend on this one (in the case that ``navigation_package`` was a library):\
\
    ``Other ROS Package`` → ``navigation_package`` → ``yaml-cpp Conan Package``\
\
We can use the ament helper `ament_export_dependencies()` to export the Conan targets as we would do with a normal ROS package. You can read more about it in our documentation: <https://docs.conan.io/2/integrations/ros.html>

Now we install the ``yaml-cpp`` Conan package like so:

```sh

> $ conan install navigation_package/conanfile.txt --build missing --output-folder install/conan

Reviewer 3: ```suggestion
$ conan install navigation_package/conanfile.txt --build=missing --output-folder=install/conan
```

We can build our ROS package as usual, just source the conanrosenv.sh file before. This will set the environment so CMake can find the files generated by Conan.

```sh
$ source install/conan/conanrosenv.sh
$ colcon build --packages-select navigation_package
Starting >>> navigation_package
Finished <<< navigation_package [16.3s]

Summary: 1 package finished [16.7s]
```

Now it is finally time to run our executable:

```sh
$ source install/setup.bash
$ ros2 run navigation_package navigator navigation_package/locations.yaml
[INFO] [1732633293.207085200] [yaml_navigation_node]: Reading locations from YAML...
[INFO] [1732633293.208468700] [yaml_navigation_node]: Sending Kitchen goal (3.50, 2.00) to robot
[INFO] [1732633293.208949200] [yaml_navigation_node]: Sending Living Room goal (1.00, -1.00) to robot
[INFO] [1732633293.209244600] [yaml_navigation_node]: Sending Bedroom goal (-2.00, 1.50) to robot
[INFO] [1732633293.209548300] [yaml_navigation_node]: All goals have been sent.
```

## Conclusion

The new ROS integration for Conan offers a neat way of incorporating Conan packages to ROS packages. By simply including a conanfile, you can

> install the required libraries from Conan Center.

Reviewer 3: ```suggestion
install the required libraries from Conan Center or from your own private server.
```
